File: LeetCode2021-2022/ts/257.binaryTreePaths.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

const dfs = (root: TreeNode | null, combina:string[], path:string):void =>{
	if(!root.left && !root.right){
		combina.push(path);
		return;
	}

	if(root.left){
		const leftP:string = path + "->" + root.left.val.toString();
		dfs(root.left, combina, leftP);
	}

	if(root.right){
		const rightP:string = path + "->" + root.right.val.toString();
		dfs(root.right, combina, rightP);
	}

	return;
}
const binaryTreePaths = (root: TreeNode | null) :string[] =>{
	if(!root){
		return [];
	}
	
	let combina:string = [];
	dfs(root, combina, root.val.toString()); 
	return;
}

File: LeetCode2021-2022/python/257.binary_tree_paths.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def binaryTreePaths(self, root: Optional[TreeNode]) -> List[str]:
        if not root:
            return []
        
        output = []
        def dfs(node, p:str) -> None:
            if not node.left and not node.right:
                output.append(p)
                return
            if node.left:
                dfs(node.left, p + "->" + str(node.left.val))
            if node.right:
                dfs(node.right, p+ "->" +str(node.right.val))
            return
        dfs(root, str(root.val))
        return output
        
                
        
